src/lib/Lower.hs

```diff
@@ -305,34 +308,61 @@ instance SinkableE (VSubstValC c) where
 instance PrettyPrec (VSubstValC c n) where
   prettyPrec (VVal s atom) = atPrec LowestPrec $ "@" <> viaShow s <+> pretty atom
 
-type TopVectorizeM = BuilderT SimpIR (ReaderT Word32 HardFailM)
-
-vectorizeLoops :: EnvReader m => Word32 -> SBlock n -> m n (SBlock n)
-vectorizeLoops vectorByteWidth (Block _ decls ans) = do
-  liftM (runHardFail . (`runReaderT` vectorByteWidth)) $ liftBuilderT $ buildBlock $ do
-    s <- vectorizeLoopsRec emptyInFrag decls
-    applySubst s ans
+-- TODO: `WriterT Errs` should suffice, no need for full `StateT`.
+type TopVectorizeM = BuilderT SimpIR (ReaderT Word32 (StateT Errs SoftResultM))
+
+vectorizeLoops :: (MonadIO (m n), MonadLogger [Output] (m n), EnvReader m)
+               => Word32 -> Abs (Binder SimpIR) SBlock n
+               -> m n (Abs (Binder SimpIR) SBlock n, Errs)
+vectorizeLoops vectorByteWidth abs = liftEnvReaderM do
+  refreshAbs abs \d (Block _ decls ans) -> do
+    (block', errs) <- liftM (runSoftResultM' . (`runStateT` mempty) . (`runReaderT` vectorByteWidth)) $ liftBuilderT $ buildBlock do
```

Comment:
Please break this line. It's incredibly long at this point

Reply:
Done. PTAL.